8365991: AArch64: Ignore BlockZeroingLowLimit when UseBlockZeroing is false

cnqpzhang opened the pull request:
Issue: 
In AArch64 port, `UseBlockZeroing` is by default set to true and `BlockZeroingLowLimit` is initialized to 256. If `DC ZVA` is supported, `BlockZeroingLowLimit` is later updated to `4 * VM_Version::zva_length()`. When `UseBlockZeroing` is set to false, all related conditional checks should ignore `BlockZeroingLowLimit`. However, the function `MacroAssembler::zero_words(Register base, uint64_t cnt)` still evaluates the lower limit and bases its code generation logic on it, which seems to be an incomplete conditional check.

This PR:
1. Reset `BlockZeroingLowLimit` to `4 * VM_Version::zva_length()` or 256 with a warning message if it was manually configured from the default while `UseBlockZeroing` is disabled.
2. Added necessary comments in `MacroAssembler::zero_words(Register base, uint64_t cnt)` and `MacroAssembler::zero_words(Register ptr, Register cnt)` to explain why we do not check `UseBlockZeroing` in the outer part of these functions. Instead, the decision is delegated to the stub function `zero_blocks`, which encapsulates the DC ZVA instructions and serves as the inner implementation of `zero_words`. This approach helps better control the increase in code cache size during array or object instance initialization.
3. Added more testing sizes to `test/micro/org/openjdk/bench/vm/gc/RawAllocationRate.java` to better cover scenarios involving smaller arrays and objects.. 

Tests:
1. Performance tests on the bundled JMH `vm.compiler.ClearMemory`, and `vm.gc.RawAllocationRate` (including `arrayTest` and `instanceTest`) showed no obvious regression. Negative tests with `jdk/bin/java -jar images/test/micro/benchmarks.jar RawAllocationRate.arrayTest_C1 -bm thrpt -gc false -wi 0 -w 30 -i 1 -r 30 -t 1 -f 1 -tu s -jvmArgs "-XX:-UseBlockZeroing -XX:BlockZeroingLowLimit=8" -p size=32` demonstrated good wall times on `zero_words_reg_imm` calls, as expected.
2. Jtreg ter1 test on Ampere Altra, AmpereOne, Graviton2 and 3, tier2 on Altra. No new issues found. Passed tests of GHA Sanity Checks.

---------
- [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8365991](https://bugs.openjdk.org/browse/JDK-8365991): AArch64: Ignore BlockZeroingLowLimit when UseBlockZeroing is false (**Enhancement** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/26917/head:pull/26917` \
`$ git checkout pull/26917`

Update a local copy of the PR: \
`$ git checkout pull/26917` \
`$ git pull https://git.openjdk.org/jdk.git pull/26917/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 26917`

View PR using the GUI difftool: \
`$ git pr show -t 26917`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/26917.diff">https://git.openjdk.org/jdk/pull/26917.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/26917#issuecomment-3218220987)
</details>

cnqpzhang committed:
8365991: AArch64: Ignore BlockZeroingLowLimit when UseBlockZeroing is false

Signed-off-by: Patrick Zhang <patrick@os.amperecomputing.com>

adinn commented:
@cnqpzhang If you look back at the history of this code you will see that you are undoing a change that was made deliberately by @theRealAph. Your patch may improve the specific test case you have provided but at the cost of a significant and unacceptable increase in code cache use for all cases.

The comment at the head of the code you have edited makes this point explicitly. The reasoning behind that comment is available in the JIRA history and associated review comments. The relevant issue is

  https://bugs.openjdk.org/browse/JDK-8179444

and the corresponding review thread starts with 

  https://mail.openjdk.org/pipermail/hotspot-dev/2017-April/026742.html

and continues with

  https://mail.openjdk.org/pipermail/hotspot-dev/2017-May/026766.html

I don't recommend integrating this change.

cnqpzhang committed:
Roll back main changes on zero_words_reg_reg and generate_zero_blocks

Signed-off-by: Patrick Zhang <patrick@os.amperecomputing.com>

cnqpzhang commented:
Regarding the impact to code caches, I measured JMH `vm.gc.RawAllocationRate.arrayTest` and `SPECjbb2015 PRESET run`. The first is not suitable for comparison because the array init code only takes a small portion of the overall space, with `-XX:+TieredCompilation` the sum of three segmented caches only showed <<1% diff. In another viewpoint, `SPECjbb2015` can be a complicated enough app that is able to demonstrate the impact on code caches, so I plot such a chart for a 20 minutes run, baseline vs patched. 

<img width="2007" height="1102" alt="image" src="https://github.com/user-attachments/assets/373a30eb-4892-4927-94af-78419fd08fb2" />

We could eyeball that the profiled and non-profiled nmethods have slightly bigger sizes of used caches (patched vs baseline), tiny part of the total sizes ~6MB (profiled nm) and ~12MB (non-profiled nm). Furthermore, these diffs are relatively far smaller than the total reserved size, either 32M (C1 only), or 48M (with C2), or 240M (configured ergonomically by JVM). I manually set it as `-XX:InitialCodeCacheSize=32M -XX:ReservedCodeCacheSize=64M` for a managed range.

Therefore, I have a question regarding the practical impact of the code cache in this context. Specifically, is the code cache still practically a significant concern relative to the benefits gained from reduced call counts and the modest performance improvements in code generation and execution for the generated array and object initialization code?

That said, I fully understand the potential risks and concerns associated with modifying the existing logic. I would get prepared to roll back the changes related to the C2 part.

cnqpzhang committed:
Reset BlockZeroingLowLimit to 4 * _zva_length

Signed-off-by: Patrick Zhang <patrick@os.amperecomputing.com>

cnqpzhang commented:
> To begin with, please add this short patch, then see if any of this PR provides an advantage.
> 
> ```
> 
> diff --git a/src/hotspot/cpu/aarch64/vm_version_aarch64.cpp b/src/hotspot/cpu/aarch64/vm_version_aarch64.cpp
> index 9321dd0542e..14a584c5106 100644
> --- a/src/hotspot/cpu/aarch64/vm_version_aarch64.cpp
> +++ b/src/hotspot/cpu/aarch64/vm_version_aarch64.cpp
> @@ -446,6 +446,11 @@ void VM_Version::initialize() {
>      FLAG_SET_DEFAULT(UseBlockZeroing, false);
>    }
>  
> +  if (!UseBlockZeroing && !FLAG_IS_DEFAULT(BlockZeroingLowLimit)) {
> +    warning("BlockZeroingLowLimit has been ignored because UseBlockZeroing is disabled");
> +    FLAG_SET_DEFAULT(BlockZeroingLowLimit, 4 * VM_Version::zva_length());
> +  }
> +
>    if (VM_Version::supports_sve2()) {
>      if (FLAG_IS_DEFAULT(UseSVE)) {
>        FLAG_SET_DEFAULT(UseSVE, 2);
> ```

Thanks for advice. Updated accordingly (commit 3 vs 2: https://github.com/openjdk/jdk/pull/26917/commits/22e72f49e82cc7febbcd44b01636f8cbdeaa4cca) to keep the shape of the generated code as unchanged as possible. My test case with `-XX:-UseBlockZeroing -XX:BlockZeroingLowLimit=8, size=32` also works as expected. I added some comments to better clarify the purpose of the if-condition inside the `zero_words` function to avoid future confusion upon. Please help review, thanks.

theRealAph commented:
> Please help review, thanks.

OK, but please edit the claims at the top of this PR to respect the new reality. In particular, please state the test cases which are improved.

cnqpzhang committed:
Check is_zva_enabled when resetting BlockZeroingLowLimit

Signed-off-by: Patrick Zhang <patrick@os.amperecomputing.com>

theRealAph commented:
I can't see any statistically-significant improvement. Please tell us your test results and your test conditions.

Commit:
Benchmark MacroAssembler::zero_words calls

Signed-off-by: Patrick Zhang <patrick@os.amperecomputing.com>

cnqpzhang committed:
fixed format string to %zu for size_t unsigned size type

Signed-off-by: Patrick Zhang <patrick@os.amperecomputing.com>

cnqpzhang committed:
Refine the count types to pass mac and win builds

Signed-off-by: Patrick Zhang <patrick@os.amperecomputing.com>

cnqpzhang commented:
Added [test/hotspot/gtest/aarch64/test_MacroAssembler_zero_words.cpp](https://github.com/openjdk/jdk/pull/26917/files#diff-6698530e8c1fe55cec7c20ca36725bde034b3cd74b25ccee77cc50f66dabf16c) to measure the impact of different low limits and cleared word counts on the wall time of `MacroAssembler::zero_words` and compare the resulting differences.

Run the test and compare the wall times. We can see that `fixing the low limit from a lower value to the default 256` improves codegen efficiency, by 11x on clear_4_words (289 vs. 25) and by 1.6x on clear_16_words (170 vs. 107).

$ make run-test TEST="gtest:MacroAssemblerZeroWordsTest"

> Clear 4 words with lower limit 8, zero_words wall time (ns): 289
> Clear 4 words with lower limit 256, zero_words wall time (ns): 25
> Clear 16 words with lower limit 64, zero_words wall time (ns): 170
> Clear 16 words with lower limit 256, zero_words wall time (ns): 107

See below for the detailed run log, including the generated code sequences under various conditions:

```
Test selection 'gtest:MacroAssemblerZeroWordsTest', will run:
* gtest:MacroAssemblerZeroWordsTest/server

Running test 'gtest:MacroAssemblerZeroWordsTest/server'
Note: Google Test filter = MacroAssemblerZeroWordsTest*
[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from MacroAssemblerZeroWordsTest
[ RUN      ] MacroAssemblerZeroWordsTest.UseBZ_clear_32B_with_lowlimit_8B_vm
--------------------------------------------------------------------------------
udf     #0
  0x0000400011c56a40:   mov     x10, #0x6fb0                    // #28592
  0x0000400011c56a44:   movk    x10, #0xab05, lsl #16
  0x0000400011c56a48:   movk    x10, #0xaaaa, lsl #32
  0x0000400011c56a4c:   orr     x11, xzr, #0x4
  0x0000400011c56a50:   subs    x8, x11, #0x8
  0x0000400011c56a54:   b.cc    0x0000400011c56a5c  // b.lo, b.ul, b.last
  0x0000400011c56a58:   bl      Stub::Stub Generator zero_blocks_stub
  0x0000400011c56a5c:   tbz     w11, #2, 0x0000400011c56a68
  0x0000400011c56a60:   stp     xzr, xzr, [x10], #16
  0x0000400011c56a64:   stp     xzr, xzr, [x10], #16
  0x0000400011c56a68:   tbz     w11, #1, 0x0000400011c56a70
  0x0000400011c56a6c:   stp     xzr, xzr, [x10], #16
  0x0000400011c56a70:   tbz     w11, #0, 0x0000400011c56a78
  0x0000400011c56a74:   str     xzr, [x10]
--------------------------------------------------------------------------------

Clear 4 words with lower limit 8, zero_words wall time (ns): 289
[       OK ] MacroAssemblerZeroWordsTest.UseBZ_clear_32B_with_lowlimit_8B_vm (2 ms)
[ RUN      ] MacroAssemblerZeroWordsTest.UseBZ_clear_32B_with_lowlimit_256B_vm
--------------------------------------------------------------------------------
udf     #0
  0x0000400011c57400:   mov     x10, #0x6fb0                    // #28592
  0x0000400011c57404:   movk    x10, #0xab05, lsl #16
  0x0000400011c57408:   movk    x10, #0xaaaa, lsl #32
  0x0000400011c5740c:   stp     xzr, xzr, [x10]
  0x0000400011c57410:   stp     xzr, xzr, [x10, #16]
--------------------------------------------------------------------------------

Clear 4 words with lower limit 256, zero_words wall time (ns): 25
[       OK ] MacroAssemblerZeroWordsTest.UseBZ_clear_32B_with_lowlimit_256B_vm (0 ms)
[ RUN      ] MacroAssemblerZeroWordsTest.UseBZ_clear_128B_with_lowlimit_64B_vm
--------------------------------------------------------------------------------
udf     #0
  0x0000400011c57400:   mov     x10, #0x6fe0                    // #28640
  0x0000400011c57404:   movk    x10, #0xab05, lsl #16
  0x0000400011c57408:   movk    x10, #0xaaaa, lsl #32
  0x0000400011c5740c:   orr     x11, xzr, #0x10
  0x0000400011c57410:   subs    x8, x11, #0x8
  0x0000400011c57414:   b.cc    0x0000400011c5741c  // b.lo, b.ul, b.last
  0x0000400011c57418:   bl      Stub::Stub Generator zero_blocks_stub
  0x0000400011c5741c:   tbz     w11, #2, 0x0000400011c57428
  0x0000400011c57420:   stp     xzr, xzr, [x10], #16
  0x0000400011c57424:   stp     xzr, xzr, [x10], #16
  0x0000400011c57428:   tbz     w11, #1, 0x0000400011c57430
  0x0000400011c5742c:   stp     xzr, xzr, [x10], #16
  0x0000400011c57430:   tbz     w11, #0, 0x0000400011c57438
  0x0000400011c57434:   str     xzr, [x10]
--------------------------------------------------------------------------------

Clear 16 words with lower limit 64, zero_words wall time (ns): 170
[       OK ] MacroAssemblerZeroWordsTest.UseBZ_clear_128B_with_lowlimit_64B_vm (0 ms)
[ RUN      ] MacroAssemblerZeroWordsTest.UseBZ_clear_128B_with_lowlimit_256B_vm
--------------------------------------------------------------------------------
udf     #0
  0x0000400011c57400:   mov     x10, #0x6fe0                    // #28640
  0x0000400011c57404:   movk    x10, #0xab05, lsl #16
  0x0000400011c57408:   movk    x10, #0xaaaa, lsl #32
  0x0000400011c5740c:   stp     xzr, xzr, [x10]
  0x0000400011c57410:   stp     xzr, xzr, [x10, #16]
  0x0000400011c57414:   stp     xzr, xzr, [x10, #32]
  0x0000400011c57418:   stp     xzr, xzr, [x10, #48]
  0x0000400011c5741c:   stp     xzr, xzr, [x10, #64]
  0x0000400011c57420:   stp     xzr, xzr, [x10, #80]
  0x0000400011c57424:   stp     xzr, xzr, [x10, #96]
  0x0000400011c57428:   stp     xzr, xzr, [x10, #112]
  0x0000400011c5742c:   add     x10, x10, #0x80
--------------------------------------------------------------------------------

Clear 16 words with lower limit 256, zero_words wall time (ns): 107
[       OK ] MacroAssemblerZeroWordsTest.UseBZ_clear_128B_with_lowlimit_256B_vm (0 ms)
[----------] 4 tests from MacroAssemblerZeroWordsTest (110 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test suite ran. (110 ms total)
[  PASSED  ] 4 tests.
Finished running test 'gtest:MacroAssemblerZeroWordsTest/server'
Test report is stored in build-pr/test-results/gtest_MacroAssemblerZeroWordsTest_server

==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR  SKIP   
   gtest:MacroAssemblerZeroWordsTest/server              4     4     0     0     0   
==============================
TEST SUCCESS

```

adinn commented:
@cnqpzhang I don't understand why you think these tests indicate anything useful for real use cases. Do you have an actual user whose needs justify adopting this change?

Let's consider what your patch and associated test achieve. Initially you tried to remove the limit on unrolling that was imposed to avoid excessive cache consumption. When it was explained why this was inappropriate you reduced the patch so that it now adjusts the threshold at which unrolling is replaced by a call to the stub. Your two test runs appear to demonstrate a performance improvement between old and new but the difference is more apparent than real. In the specific configurations you have selected your change to the unrolling threshold targets two very specific points of disparity. The new cases fully unroll while the old cases rely on a callout. Not surpisingly. this gives very different performance when you run it in a loop many times. But we already know that callouts are more expensive than inline code.

The important thing to note is that this transition between unrolling vs callout happens in both old and new code, just at different size points. If you ran with other config settings and sizes you could find many cases where both versions fully unroll or both rely on a callout. So your test does not truly reflect what is going on here and your fix is really doing little more than rescaling the dials so they can go up to 11. You have provided no good evidence as to why we need to adjust the scale by which we compute the threshold between unrolling or callout. Furthermore, since this rescaling allows more unrolling to occur than in the old version you still need to justify why that is worth doing.

theRealAph commented:
> > > I would like to reiterate that I have no objection to the functions when the `-XX:+UseBlockZeroing` option is set, everything can keep as is. My point is that `BlockZeroingLowLimit` serves literally/specifically as a switch to control whether DC ZVA instructions are generated for clearing instances under a specified bytes size limitation, rather than for deciding between unrolling and callout. Therefore, it should NOT affect the code-gen results any longer when `-XX:-UseBlockZeroing` is set, should it?
> > 
> > 
> > It does not. When `-XX:-UseBlockZeroing` is set, `BlockZeroingLowLimit` is ignored.
> 
> zero_words does not check `UseBlockZeroing`, it directly compares `cnt` and `BlockZeroingLowLimit / BytesPerWord`.

It doesn't need to because

```
  if (!UseBlockZeroing && !FLAG_IS_DEFAULT(BlockZeroingLowLimit)) {
    warning("BlockZeroingLowLimit has been ignored because UseBlockZeroing is disabled");
    FLAG_SET_DEFAULT(BlockZeroingLowLimit, is_zva_enabled() ? (4 * VM_Version::zva_length()) : 256);
  }
```

That is to say, if a user sets `BlockZeroingLowLimit` and `-XX:-UseBlockZeroing`, then the user's `BlockZeroingLowLimit` is, rightly, ignored.

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -6203,8 +6203,6 @@ address MacroAssembler::zero_words(Register ptr, Register cnt)
   BLOCK_COMMENT("zero_words {");
   assert(ptr == r10 && cnt == r11, "mismatch in register usage");
-  RuntimeAddress zero_blocks = RuntimeAddress(StubRoutines::aarch64::zero_blocks());
-  assert(zero_blocks.target() != nullptr, "zero_blocks stub has not been generated");
```
What is the point of this change?

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -6259,47 +6259,55 @@ address MacroAssembler::zero_words(Register ptr, Register cnt)
+    mov(r10, base); mov(r11, cnt);
+    result = zero_words(r10, r11);
+  } else {
 #ifndef PRODUCT
```
What is this change for?

cnqpzhang commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -6259,47 +6259,55 @@ address MacroAssembler::zero_words(Register ptr, Register cnt)
+  // zero_blocks stub function that wraps the core logic of zero_words
+  // and necessary unrolled str/stp expanding when the condition is not met.
+  // This approach also helps prevent sudden increases in code cache size
+  // when zeroing large memory areas in many places.
```
Updated

cnqpzhang commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -6259,47 +6259,55 @@ address MacroAssembler::zero_words(Register ptr, Register cnt)
-    if (cnt >= 16) {
-      uint64_t loops = cnt/16;
+    // Use 16 words as the block size which is 128 bytes on 64-bit systems.
+    // A complete loop body will be 8 STPs unrolled there.
```
Updated

cnqpzhang reviewed (commented):
Thanks for review, please see my updates and replies with the new commit.

cnqpzhang committed:
Improve the comments for zero_words funcs

Signed-off-by: Patrick Zhang <patrick@os.amperecomputing.com>

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -6259,47 +6259,55 @@ address MacroAssembler::zero_words(Register ptr, Register cnt)
-      uint64_t loops = cnt/16;
+    // Use 16 words as the block size which is 128 bytes on 64-bit systems.
+    // A complete loop body will be 8 STPs unrolled there.
+    const int block_size = 16;
```
> I wondered why `MacroAssembler::zero_words` uses 16 words to do `stp` unrolling, while `generate_zero_blocks()` 8 words (`const int MacroAssembler::zero_words_block_size = 8;`), so defined this variable to compare `8 vs 16` but did not find obvious performance difference.
> 
> Regarding the var name `block_size`, could `unroll` or `unroll_words` be better?

What's wrong with 16? I'm asking not from a "my teachers said always name constants" point of view, but from a reader's understanding point of view. Named constants are all well and good if the constant has some meaning, but this one is just two words. Perhaps `2 * WordSize` would do.

cnqpzhang replied:
[PR 4919](https://github.com/openjdk/jdk/pull/4919/files#diff-0f4150a9c607ccd590bf256daa800c0276144682a92bc6bdced5e8bc1bb81f3aL4737) for [8270947: AArch64: C1: use zero_words to initialize all objects](https://bugs.openjdk.org/browse/JDK-8270947) changed the loop unroll length from 4 STPs (clearing 8 words) to 8 STPs (clearing 16 words) within the loop. In fact, before that PR, it seems the value 16 was from the `SmallArraySize` (`18 * BytesPerLong`).

For the current code snippet, I think, using `2 * WordSize` to the constant is not ideal, as it suggests the size of the unroll length and can be confusing. A better choice might be `2 * MacroAssembler::zero_words_block_size`, which more accurately reflects the intended meaning. Is this acceptable?

cnqpzhang replied:
Please see the latest commit. I renamed the var name to `unroll_words` in `address MacroAssembler::zero_words(Register base, uint64_t cnt)` and `address generate_zero_blocks()` and have them aligned, one for 16 words and the other for 8 words to do unrolling for block clearing.

cnqpzhang committed:
Changed the constant 16 to 2 * zero_words_block_size

Signed-off-by: Patrick Zhang <patrick@os.amperecomputing.com>

cnqpzhang committed:
Renamed the vars to unroll_words

Signed-off-by: Patrick Zhang <patrick@os.amperecomputing.com>

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -6259,47 +6257,52 @@ address MacroAssembler::zero_words(Register ptr, Register cnt)
-  if (cnt <= (uint64_t)BlockZeroingLowLimit / BytesPerWord) {
+  // There is no need to check UseBlockZeroing here because that is
+  // delegated to the zero_blocks stub. The code here is inlined, so
+  // it is important to keep it small.
```
```suggestion
  // The code here is inlined, so it is important to keep it small.
```
I find this "There is no need..." comment more confusing than helpful.

cnqpzhang committed:
Rename unroll_words to two_blocks

Signed-off-by: Patrick Zhang <xmldk@tuta.io>

cnqpzhang committed:
Trigger OCA recheck
